File: java/src/main/java/io/valkey/javabenchmark/metrics/NdjsonMetricsWriter.java
/*
 * Copyright 2025 the original author or authors.
 */
package io.valkey.javabenchmark.metrics;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.valkey.javabenchmark.metrics.MetricsCollector.CommandMetrics;
import org.HdrHistogram.Histogram;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.nio.ByteBuffer;
import java.nio.file.*;
import java.time.Instant;
import java.util.Base64;
import java.util.Map;

/**
 * Writes benchmark metrics to NDJSON format (Newline Delimited JSON).
 * Each phase is written as a single JSON line, allowing the orchestrator
 * to detect phase completion by watching for new lines.
 * 
 * Includes native HDR histogram encoding with base64 payload for full fidelity latency data.
 *
 * @author Ilia Kolominsky
 */
public class NdjsonMetricsWriter {
    private static final Logger logger = LoggerFactory.getLogger(NdjsonMetricsWriter.class);
    
    private final Path outputPath;
    private final ObjectMapper objectMapper;

    public NdjsonMetricsWriter(String path) {
        this.outputPath = Path.of(path);
        this.objectMapper = new ObjectMapper();
        // No pretty printing - NDJSON requires compact single-line JSON
    }

    /**
     * Writes phase results as a single NDJSON line.
     */
    public void writePhaseResults(String phaseId, String status, int connections, 
                                   MetricsCollector collector) throws IOException {
        
        // Ensure parent directory exists
        Files.createDirectories(outputPath.getParent());
        
        // Build the JSON object
        ObjectNode root = buildPhaseJson(phaseId, status, connections, collector);
        
        // Write as single line (no pretty printing) + newline
        try (BufferedWriter writer = Files.newBufferedWriter(outputPath, 
                StandardOpenOption.CREATE, StandardOpenOption.APPEND)) {
            writer.write(objectMapper.writeValueAsString(root));
            writer.newLine();
        }
        
        logger.info("Wrote NDJSON metrics for phase {} to {}", phaseId, outputPath);
    }

    private ObjectNode buildPhaseJson(String phaseId, String status, int connections,
                                       MetricsCollector collector) {
        ObjectNode root = objectMapper.createObjectNode();
        
        // Phase metadata
        ObjectNode phase = root.putObject("phase");
        phase.put("id", phaseId);
        phase.put("status", status);
        phase.put("start_timestamp", Instant.ofEpochMilli(collector.getStartTime()).toString());
        phase.put("finish_timestamp", Instant.ofEpochMilli(collector.getEndTime()).toString());
        phase.put("duration_ms", collector.getDurationMillis());
        phase.put("connections", connections);
        
        // Totals
        ObjectNode totals = root.putObject("totals");
        totals.put("requests", collector.getTotalRequests());
        totals.put("errors", collector.getTotalErrors());
        
        // Per-command metrics
        ObjectNode metrics = root.putObject("metrics");
        for (Map.Entry<String, CommandMetrics> entry : collector.getAllMetrics().entrySet()) {
            String cmdName = entry.getKey();
            CommandMetrics cmdMetrics = entry.getValue();
            
            ObjectNode cmdNode = metrics.putObject(cmdName);
            cmdNode.put("requests", cmdMetrics.getRequests());
            cmdNode.put("errors", cmdMetrics.getErrors());
            
            // Latency sub-object
            ObjectNode latency = cmdNode.putObject("latency");
            latency.put("unit", "us");
            latency.put("count", cmdMetrics.getHistogram().getTotalCount());
            
            // Summary percentiles
            ObjectNode summary = latency.putObject("summary");
            summary.put("min", cmdMetrics.getMin());
            summary.put("p50", cmdMetrics.getP50());
            summary.put("p95", cmdMetrics.getP95());
            summary.put("p99", cmdMetrics.getP99());
            summary.put("p999", cmdMetrics.getP999());
            summary.put("max", cmdMetrics.getMax());
            
            // HDR histogram with native encoding
            ObjectNode hdr = latency.putObject("hdr");
            hdr.put("format", "hdr");
            hdr.put("sigfig", 3);
            hdr.put("payload_b64", encodeHistogram(cmdMetrics.getHistogram()));
        }
        
        return root;
    }

    /**
     * Encodes the histogram to a base64 string using HDR's native compressed format.
     */
    private String encodeHistogram(Histogram histogram) {
        try {
            int neededCapacity = histogram.getNeededByteBufferCapacity();
            ByteBuffer buffer = ByteBuffer.allocate(neededCapacity);
            int bytesWritten = histogram.encodeIntoCompressedByteBuffer(buffer);
            
            // Extract only the bytes that were written
            byte[] compressedBytes = new byte[bytesWritten];
            buffer.flip();
            buffer.get(compressedBytes);
            
            return Base64.getEncoder().encodeToString(compressedBytes);
        } catch (Exception e) {
            logger.warn("Failed to encode histogram: {}", e.getMessage());
            return "";
        }
    }
}

File: csharp/src/RespBench/Metrics/NdjsonMetricsWriter.cs
/*
 * Copyright 2025 the original author or authors.
 */
using System.IO;
using System.Text.Json;
using System.Text.Json.Nodes;
using HdrHistogram;
using HdrHistogram.Encoding;
using HdrHistogram.Utilities;

namespace RespBench.Metrics;

/// <summary>
/// Writes benchmark metrics to NDJSON format.
/// Includes native HDR histogram encoding with base64 payload.
/// </summary>
public class NdjsonMetricsWriter
{
    private readonly string _outputPath;

    private string? _commitId;
    private string? _driverId;
    private string? _primaryDriverVersion;
    private string? _secondaryDriverId;
    private string? _secondaryDriverVersion;

    public NdjsonMetricsWriter(string path)
    {
        _outputPath = path;
    }

    public void SetMetadata(string? commitId, string? driverId, string? primaryDriverVersion,
                           string? secondaryDriverId, string? secondaryDriverVersion)
    {
        _commitId = commitId;
        _driverId = driverId;
        _primaryDriverVersion = primaryDriverVersion;
        _secondaryDriverId = secondaryDriverId;
        _secondaryDriverVersion = secondaryDriverVersion;
    }

    public void WritePhaseResults(string phaseId, string status, int connections, MetricsCollector collector)
    {
        // Ensure parent directory exists
        var dir = Path.GetDirectoryName(_outputPath);
        if (!string.IsNullOrEmpty(dir)) Directory.CreateDirectory(dir);

        var root = BuildPhaseJson(phaseId, status, connections, collector);

        // Write as single line + newline
        string json = root.ToJsonString(new JsonSerializerOptions { WriteIndented = false });
        File.AppendAllText(_outputPath, json + Environment.NewLine);
    }

    private JsonObject BuildPhaseJson(string phaseId, string status, int connections, MetricsCollector collector)
    {
        var root = new JsonObject();

        // Metadata
        if (_commitId != null || _driverId != null)
        {
            var metadata = new JsonObject();
            if (_commitId != null) metadata["commit_id"] = _commitId;
            metadata["timestamp"] = DateTimeOffset.UtcNow.ToString("o");
            if (_driverId != null) metadata["driver_id"] = _driverId;
            if (_primaryDriverVersion != null) metadata["primary_driver_version"] = _primaryDriverVersion;
            if (_secondaryDriverId != null) metadata["secondary_driver_id"] = _secondaryDriverId;
            if (_secondaryDriverVersion != null) metadata["secondary_driver_version"] = _secondaryDriverVersion;
            root["metadata"] = metadata;
        }

        // Phase metadata
        var phase = new JsonObject
        {
            ["id"] = phaseId,
            ["status"] = status,
            ["start_timestamp"] = DateTimeOffset.FromUnixTimeMilliseconds(collector.StartTime).ToString("o"),
            ["finish_timestamp"] = DateTimeOffset.FromUnixTimeMilliseconds(collector.EndTime).ToString("o"),
            ["duration_ms"] = collector.DurationMillis,
            ["connections"] = connections
        };
        root["phase"] = phase;

        // Totals
        var totals = new JsonObject
        {
            ["requests"] = collector.TotalRequests,
            ["errors"] = collector.TotalErrors
        };
        root["totals"] = totals;

        // Per-command metrics
        var metrics = new JsonObject();
        foreach (var (cmdName, cmdMetrics) in collector.AllMetrics)
        {
            var cmdNode = new JsonObject
            {
                ["requests"] = cmdMetrics.Requests,
                ["errors"] = cmdMetrics.Errors
            };

            var latency = new JsonObject
            {
                ["unit"] = "us",
                ["count"] = cmdMetrics.Histogram.TotalCount
            };

            var summary = new JsonObject
            {
                ["min"] = cmdMetrics.Histogram.TotalCount > 0 ? cmdMetrics.Histogram.GetValueAtPercentile(0) : 0,
                ["p50"] = cmdMetrics.P50,
                ["p95"] = cmdMetrics.P95,
                ["p99"] = cmdMetrics.P99,
                ["p999"] = cmdMetrics.P999,
                ["max"] = cmdMetrics.Max
            };
            latency["summary"] = summary;

            var hdr = new JsonObject
            {
                ["format"] = "hdr",
                ["sigfig"] = 3,
                ["payload_b64"] = EncodeHistogram(cmdMetrics.Histogram)
            };
            latency["hdr"] = hdr;

            cmdNode["latency"] = latency;
            metrics[cmdName] = cmdNode;
        }
        root["metrics"] = metrics;

        return root;
    }

    private static string EncodeHistogram(HistogramBase histogram)
    {
        try
        {
            // Use HdrHistogram.NET's direct encoding API (matches Java's encodeIntoCompressedByteBuffer)
            int neededCapacity = histogram.GetNeededByteBufferCapacity();
            var buffer = ByteBuffer.Allocate(neededCapacity);
            int bytesWritten = histogram.Encode(buffer, HistogramEncoderV2.Instance);

            // Extract the written bytes: reset position, read one at a time
            buffer.Position = 0;
            byte[] compressedBytes = new byte[bytesWritten];
            for (int i = 0; i < bytesWritten; i++)
                compressedBytes[i] = buffer.Get();

            return Convert.ToBase64String(compressedBytes);
        }
        catch
        {
            return "";
        }
    }
}
